### ISSUE TITLE
@reduxjs/toolkit doesn't sit well with react-redux 9.0.0+

### ISSUE THREAD

**Issue by VincentLu91** [reduxjs/redux-toolkit#3959]
No matter the version of @reduxjs/toolkit (in my case, either 2.0.1. or 1.9.7), I can't install the new react-redux.

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: next-app@undefined
npm ERR! Found: react-redux@8.1.3
npm ERR! node_modules/react-redux
npm ERR!   react-redux@"^9.0.0" from the root project
npm ERR!   peerOptional react-redux@"^7.2.1 || ^8.1.3 || ^9.0.0" from @reduxjs/toolkit@2.0.1
npm ERR!   node_modules/@reduxjs/toolkit
npm ERR!     @reduxjs/toolkit@"^2.0.1" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! react-redux@"^9.0.0" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: redux@5.0.0
npm ERR! node_modules/redux
npm ERR!   peerOptional redux@"^5.0.0" from react-redux@9.0.2
npm ERR!   node_modules/react-redux
npm ERR!     react-redux@"^9.0.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! 
npm ERR! For a full report see:
npm ERR! /Users/vince/.npm/_logs/2023-12-06T23_10_06_622Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in: /Users/vince/.npm/_logs/2023-12-06T23_10_06_622Z-debug-0.log
```

**Comment by phryneas** [comment]
So which versions of `react-redux`, `redux` and `@reduxjs/toolkit` do you have installed? It should be 9.x.x,5.x.x, 2.x.x

**Comment by markerikson** [comment]
Hmm. I just tried running that myself in a new empty project with NPM 9.8.1, and it worked.

What version of NPM are you using? Is this in an empty project or an existing one?

**Comment by VincentLu91** [comment]
> Hmm. I just tried running that myself in a new empty project with NPM 9.8.1, and it worked.
> 
> What version of NPM are you using? Is this in an empty project or an existing one?

npm version: 9.8.1
node version: v18.18.2

I tried it on both my existing app and a new one. 

I just started a new project to prove this is an issue:
```npx create-next-app --example with-redux nextjs_redux_new```

The dependencies:

```"dependencies": {
    "@reduxjs/toolkit": "1.9.5",
    "next": "latest",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-redux": "8.1.0"
  },```

Then when I installed the newest version of react-redux, it still fails. Even after upgrading reduxjs/toolkit

```
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR! 
npm ERR! While resolving: undefined@undefined
npm ERR! Found: react-redux@8.1.0
npm ERR! node_modules/react-redux
npm ERR!   react-redux@"9.0.0" from the root project
npm ERR!   peerOptional react-redux@"^7.2.1 || ^8.0.2" from @reduxjs/toolkit@1.9.5
npm ERR!   node_modules/@reduxjs/toolkit
npm ERR!     @reduxjs/toolkit@"1.9.5" from the root project
npm ERR!     peerOptional @reduxjs/toolkit@"^1 || ^2.0.0-beta.0" from react-redux@8.1.0
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! react-redux@"9.0.0" from the root project
npm ERR! 
npm ERR! Conflicting peer dependency: redux@5.0.0
npm ERR! node_modules/redux
npm ERR!   peerOptional redux@"^5.0.0-rc" from react-redux@9.0.0
npm ERR!   node_modules/react-redux
npm ERR!     react-redux@"9.0.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! 
npm ERR! For a full report see:
npm ERR! /Users/vince/.npm/_logs/2023-12-07T17_00_12_993Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in: /Users/vince/.npm/_logs/2023-12-07T17_00_12_993Z-debug-0.log
(base) Vincents-MBP:nextjs_redux_new vincentlu$ 
```

**Comment by markerikson** [comment]
Ah. React-Redux 9.0.0 had a bad `redux` dependency. You need to install 9.0.2, which is the latest and has a fix for that.

**Comment by VincentLu91** [comment]
Honestly, I tried 9.0.2 as well, same thing. That was the first time I experienced the error, which was why I shifted to a slightly earlier version as close to a working version (8.1.3) as possible.

I tried it with either "@reduxjs/toolkit" being 1.9.5 or 2.0.1

**Comment by markerikson** [comment]
Can you provide _exact_ steps to reproduce this issue?

**Comment by markerikson** [comment]
Okay, I do see that issue after running:

```
npx create-next-app --example with-redux nextjs_redux_new
cd nextjs_redux-new
npm add @reduxjs/toolkit@2.0.1 react-redux@9.0.2
```

Frankly, this looks like it's NPM being really stupid about peer dependencies :(

**Comment by markerikson** [comment]
It looks like the actual problem here is that the Next project defaults to listing:

```ts
"@types/react": "18.2.8",
"@types/react-dom": "18.2.5",
```

React-Redux as currently published wants newer versions:

```ts
"@types/react": "^18.2.41",
"@types/react-dom": "^18.2.17",
```

If i update the Next app's `package.json` to list those same newer React types versions, _before_ trying to upgrade React-Redux and Redux Toolkit, it installs fine:

```
$ npm add @reduxjs/toolkit@2.0.1 react-redux@9.0.2

added 1 package, removed 4 packages, changed 8 packages, and audited 44 packages in 1s

4 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

So, try that?

**Comment by VincentLu91** [comment]
it works on the new project.
however on my existing project, it still errors out. See my package.json

```
{
  "name": "my-app",
  "private": true,
  "scripts": {
    "dev": "next dev -p 3001",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@chatscope/chat-ui-kit-react": "^1.10.1",
    "@redux-devtools/extension": "^3.2.6",
    "@reduxjs/toolkit": "^1.9.5",
    "@sanity/block-content-to-react": "^3.0.0",
    "@sanity/client": "^6.9.1",
    "@sanity/image-url": "^1.0.2",
    "@stripe/stripe-js": "^2.2.0",
    "@supabase/auth-helpers-nextjs": "^0.8.7",
    "@supabase/auth-helpers-react": "^0.4.2",
    "@supabase/auth-ui-react": "^0.4.6",
    "@supabase/supabase-js": "^2.39.0",
    "axios": "^1.6.2",
    "cors": "^2.8.5",
    "express": "^4.18.2",
    "file2arraybuffer": "^1.2.2",
    "get-blob-duration": "^1.2.0",
    "moment": "^2.29.4",
    "next": "^14.0.3",
    "next-sanity-image": "^6.1.1",
    "npm-check-updates": "^16.14.11",
    "raw-body": "^2.5.2",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-media-recorder": "1.6.5",
    "react-phone-number-input": "^3.3.7",
    "react-redux": "^8.1.3",
    "react-select": "^5.8.0",
    "recordrtc": "^5.6.2",
    "redux-persist": "^6.0.0",
    "stripe": "^14.7.0",
    "styled-jsx": "^5.1.2"
  },
  "devDependencies": {
    "@types/node": "20.3.1",
    "eslint": "8.55.0",
    "eslint-config-next": "^14.0.3",
    "@types/react": "^18.2.41",
    "@types/react-dom": "^18.2.17",
    "@types/redux-logger": "3.0.9",
    "redux-logger": "3.0.6"
  }
}
```

**Comment by markerikson** [comment]
Honestly, I'm not sure if there's much I can do here. This really seems like an NPM-level problem.  (And frankly NPM's error messages are _bad_ - I'm reading that, and it's not very clear _what_ the actual problem is.)

My only other suggestion atm is to try removing the existing Redux-related deps, and then try re-adding the new versions.

**Comment by markerikson** [comment]
I will note that `redux-logger` has a peer dependency of `redux@^4`, which _may_ be part of the issue here.

(We actually mostly recommend against using `redux-logger` anyway - the Redux DevTools provide the same info, and in a better format.)

**Comment by VincentLu91** [comment]
I uninstalled redux-logger but it made no difference.

**Comment by phryneas** [comment]
The error message is confusing, but not unreadable. Here's that part you have to look for in the original message:
![grafik](https://github.com/reduxjs/redux-toolkit/assets/4282439/e0af8b81-e7d4-4413-8454-64ce1081268f)

As for why you have 

> Conflicting peer dependency: redux@5.0.0

in your latest error message, can you do `npm ls redux` and `npm why redux`?

**Comment by VincentLu91** [comment]
```
(base) Vince-MBP:project-foldervince$ npm ls redux
next-firebase-login-email@ /project-folder
├─┬ @redux-devtools/extension@3.2.6
│ └── redux@4.2.1
├─┬ @reduxjs/toolkit@2.0.1
│ ├─┬ redux-thunk@3.1.0
│ │ └── redux@5.0.0 deduped
│ └── redux@5.0.0
├─┬ react-redux@8.1.3
│ └── redux@4.2.1 deduped
└─┬ redux-persist@6.0.0
  └── redux@4.2.1 deduped

(base) Vince-MBP:project-foldervince$ npm ls redux
```


```
(base) Vince-MBP:project-foldervince$ npm why redux
redux@5.0.0
node_modules/@reduxjs/toolkit/node_modules/redux
  redux@"^5.0.0" from @reduxjs/toolkit@2.0.1
  node_modules/@reduxjs/toolkit
    @reduxjs/toolkit@"^2.0.1" from the root project
  peer redux@"^5.0.0" from redux-thunk@3.1.0
  node_modules/@reduxjs/toolkit/node_modules/redux-thunk
    redux-thunk@"^3.1.0" from @reduxjs/toolkit@2.0.1
    node_modules/@reduxjs/toolkit
      @reduxjs/toolkit@"^2.0.1" from the root project

redux@4.2.1 peer
node_modules/redux
  peer redux@"^3.1.0 || ^4.0.0" from @redux-devtools/extension@3.2.6
  node_modules/@redux-devtools/extension
    @redux-devtools/extension@"^3.2.6" from the root project
  peerOptional redux@"^4 || ^5.0.0-beta.0" from react-redux@8.1.3
  node_modules/react-redux
    react-redux@"^8.1.3" from the root project
    peerOptional react-redux@"^7.2.1 || ^8.1.3 || ^9.0.0" from @reduxjs/toolkit@2.0.1
    node_modules/@reduxjs/toolkit
      @reduxjs/toolkit@"^2.0.1" from the root project
  peer redux@">4.0.0" from redux-persist@6.0.0
  node_modules/redux-persist
    redux-persist@"^6.0.0" from the root project
(base) Vince-MBP:project-foldervince$
```

**Comment by phryneas** [comment]
I believe you can remove `@redux-devtools/extension` (the devtools should work without that).
Then you have `redux-persist` keeping you back. Until they update, you'll probably need something like `patch-package` to change their `package.json` to allow for `redux` 5. 

Or you use [package.json overrides](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#overrides) to just fix redux to v5 in general.

**Comment by VincentLu91** [comment]
uninstalled `@redux-devtools/extension`

not sure i'm following. redux-persist fixes redux to version 4? I'm not familiar with the syntax to override differences in versions. How do I do this?

**Comment by phryneas** [comment]
in your `package.json`:

```json
{
  "dependencies": { ... },
  "overrides": {
    "redux-persist": {
      "redux": "^5.0.0"
    }
  }
}
```

**Comment by VincentLu91** [comment]
that did the trick. Thanks.
so do I need to use override from now on whenever I upgrade new versions of redux libraries?

**Comment by phryneas** [comment]
Until a new version of `redux-persist` is published, yes, whenever a new redux major version is published. 
This only bumps the dependency from `redux-persist` to `redux`.

**Comment by ProPhycient5** [comment]
This is one of the major issue, even I also faced today. But anyways, I fixed this version issue by using following version on package.json.

```
 "dependencies": {
 .....
    "@reduxjs/toolkit": "^1.9.5",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "react-redux": "^8.0.5",
    "redux-persist": "^6.0.0",
    "redux-persist-indexeddb-storage": "^1.0.4"
  },
  ```

**Comment by markerikson** [comment]
@ProPhycient5 that's not really a fix for "upgrading to React-Redux v9", though :)

Did you try overriding versions to ensure your package manager _only_ has the latest Redux versions installed?

**Comment by ProPhycient5** [comment]
Yes @markerikson , I have also tried that too as you have suggested in the above thread. It didn't help.

**Comment by markerikson** [comment]
@ProPhycient5 so what _specific_ errors are you seeing when you try to upgrade atm?

**Comment by ProPhycient5** [comment]
@markerikson, although I didn't get error while installing but I have gotten it during deployment at vercel.

```
[17:55:47.290] Running build in Washington, D.C., USA (East) – iad1
[17:55:47.393] Cloning github.com/ProPhycient5/marvel-site (Branch: main, Commit: cce6e06)
[17:55:47.710] Cloning completed: 316.054ms
[17:55:47.765] Previous build cache not available
[17:55:48.053] Running "vercel build"
[17:55:48.554] Vercel CLI 32.6.1
[17:55:48.959] Installing dependencies...
[17:55:50.149] npm ERR! code ERESOLVE
[17:55:50.153] npm ERR! ERESOLVE could not resolve
[17:55:50.154] npm ERR! 
[17:55:50.154] npm ERR! While resolving: react-redux@9.0.2
[17:55:50.154] npm ERR! Found: @types/react@18.2.6
[17:55:50.154] npm ERR! node_modules/@types/react
[17:55:50.155] npm ERR!   @types/react@"18.2.6" from the root project
[17:55:50.155] npm ERR!   @types/react@"*" from @types/react-dom@18.2.4
[17:55:50.155] npm ERR!   node_modules/@types/react-dom
[17:55:50.155] npm ERR!     @types/react-dom@"18.2.4" from the root project
[17:55:50.155] npm ERR!   1 more (@types/react-scroll)
[17:55:50.155] npm ERR! 
[17:55:50.155] npm ERR! Could not resolve dependency:
[17:55:50.156] npm ERR! peerOptional @types/react@"^18.2.41" from react-redux@9.0.2
[17:55:50.156] npm ERR! node_modules/react-redux
[17:55:50.156] npm ERR!   react-redux@"^9.0.0" from the root project
[17:55:50.156] npm ERR!   peerOptional react-redux@"^7.2.1 || ^8.1.3 || ^9.0.0" from @reduxjs/toolkit@2.0.1
[17:55:50.156] npm ERR!   node_modules/@reduxjs/toolkit
[17:55:50.156] npm ERR!     @reduxjs/toolkit@"^2.0.0" from the root project
[17:55:50.156] npm ERR! 
[17:55:50.157] npm ERR! Conflicting peer dependency: @types/react@18.2.43
[17:55:50.157] npm ERR! node_modules/@types/react
[17:55:50.157] npm ERR!   peerOptional @types/react@"^18.2.41" from react-redux@9.0.2
[17:55:50.157] npm ERR!   node_modules/react-redux
[17:55:50.157] npm ERR!     react-redux@"^9.0.0" from the root project
[17:55:50.157] npm ERR!     peerOptional react-redux@"^7.2.1 || ^8.1.3 || ^9.0.0" from @reduxjs/toolkit@2.0.1
[17:55:50.157] npm ERR!     node_modules/@reduxjs/toolkit
[17:55:50.158] npm ERR!       @reduxjs/toolkit@"^2.0.0" from the root project
[17:55:50.158] npm ERR! 
[17:55:50.158] npm ERR! Fix the upstream dependency conflict, or retry
[17:55:50.158] npm ERR! this command with --force or --legacy-peer-deps
[17:55:50.158] npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
[17:55:50.158] npm ERR! 
[17:55:50.159] npm ERR! 
[17:55:50.159] npm ERR! For a full report see:
[17:55:50.159] npm ERR! /vercel/.npm/_logs/2023-12-10T12_25_49_417Z-eresolve-report.txt
[17:55:50.160] 
[17:55:50.161] npm ERR! A complete log of this run can be found in: /vercel/.npm/_logs/2023-12-10T12_25_49_417Z-debug-0.log
[17:55:50.181] Error: Command "npm install" exited with 1
[17:55:50.500] 
```

**Comment by markerikson** [comment]
@ProPhycient5 what version of `@types/react` do you have listed in your own `package.json`?

**Comment by ProPhycient5** [comment]
` "@types/react": "18.2.6",`

**Comment by markerikson** [comment]
Yeah, that's the other problem here.

I just published React-Redux 9.0.3, which lowers the React types dep. But, it still needs to be at least 18.2.25. Please try updating your project to at least that version, and preferably the latest 18.2.41.

**Comment by ProPhycient5** [comment]
Thanks @markerikson for prompt response.
